### DISCUSSION TITLE
Deprecating an enum value when using it for query parameters

### DISCUSSION THREAD

**Discussion by minhduc0711** [fastapi/fastapi#12209] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from fastapi import FastAPI, Query
from enum import StrEnum


class LLMEngine(StrEnum):
    GPT_35 = "gpt-3.5"  # <---- for example, I want to deprecate this value
    GPT_4O = "gpt-4o"
    GPT_4O_MINI = "gpt-4o-mini"


app = FastAPI()


@app.get("/")
async def root(llm: Annotated[LLMEngine, Query()]):
    return {"message": llm}
```

### Description

What is the proper way to mark a predefined value for a query parameter as deprecated? I'd expect that the deprecation would be shown on the Swagger docs as well.

### Operating System

Linux

### FastAPI Version

0.114.0

### Pydantic Version

2.9.1

### Python Version

3.11.9

**Comment by n8sty** [comment]
Try this:

```
from pydantic import Field

@app.get("/")
async def root(llm: Annotated[LLMEngine, Query()]) = Field(..., deprecated=True, description="This is deprecated"):
    return {"message": llm}
```

**Reply by minhduc0711** [reply]
This marks the query parameter itself as deprecated. I only want to deprecate a certain value that this query parameter accepts.

**Reply by PhysicallyActive** [reply]
How about the following?

```python
@app.get("/")
async def root(
    llm: Annotated[
        LLMEngine,
        Query(
            description=(
                f"{LLMEngine.GPT_35} is deprecated and will be removed in future versions."
            )
        ),
    ]
):
    return {"message": llm}
```

---

What's the goal? Do you want the deprecation to show up in the OpenAPI schema? Deprecation is technically just the indication that something is obsolete.

**Reply by minhduc0711** [reply]
> Do you want the deprecation to show up in the OpenAPI schema?

I'd expect in the swagger docs, the deprecated item will be formatted with strikethrough text or something like that. But I guess adding a description is also acceptable.

**Reply by PhysicallyActive** [reply]
> I'd expect in the swagger docs, the deprecated item will be formatted with strikethrough text or something like that. But I guess adding a description is also acceptable.

I do not know whether that's possible without directly altering the openapi schema (previously known as the swagger spec).

Using `Field(deprecated=True)` won't work because applying `Field(deprecated=True)` marks the entire parameter as deprecated, not just one of its possible variants.
I also tried solving it with Discriminated Unions before I realised that it's not possible to use DUs for query parameters.

@EDIT: There seems to be no support for what you're asking for in the [OpenAPI 3.1.0 specs](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object)

**Reply by minhduc0711** [reply]
Looks like it's still possible with OpenAPI 3.1.0: https://stackoverflow.com/a/78232576/10478661

However, I tried both of these on https://editor-next.swagger.io/, and in both cases the documentation page seems wrong. Example YAML:
```yaml
openapi: 3.1.0
info:
  title: Swagger Petstore - OpenAPI 3.1.0
  version: 0.0.0
paths:
  /findByStatus:
    get:
      tags:
        - pet
      summary: Finds Pets by status
      description: Multiple status values can be provided with comma separated strings
      operationId: findPetsByStatus
      parameters:
        - name: status
          in: query
          description: Status values that need to be considered for filter
          required: false
          explode: true
          schema:
            oneOf:
              - const: 'available'
              - const: 'pending'
              - const: 'sold'
                deprecated: true
      responses:
        '200':
          description: successful operation
        '400':
          description: Invalid status value
```

**Comment by YuriiMotov** [comment, marked as answer]
I think final deprecation of one of the Enum's values will be braking change and you will still need to create new version of your endpoint and deprecate old version of the endpoint.

So, I would suggest this approach:
```python
from typing import Annotated
from fastapi import FastAPI, Query
from enum import StrEnum


class LLMEngineDepr(StrEnum):
    GPT_35 = "gpt-3.5"
    GPT_4O = "gpt-4o"
    GPT_4O_MINI = "gpt-4o-mini"


class LLMEngine(StrEnum):
    GPT_4O = "gpt-4o"
    GPT_4O_MINI = "gpt-4o-mini"


app = FastAPI()


@app.get("/", deprecated=True)
async def root(llm: Annotated[LLMEngineDepr, Query()]):
    return {"message": llm}


@app.get("/v2/")
async def root_v2(llm: Annotated[LLMEngine, Query()]):
    return await root(llm=llm)

```
